virtcontainers/ich.go

```diff
@@ -0,0 +1,813 @@
+// Copyright (c) 2019 Ericsson Eurolab Deutschland GmbH
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package virtcontainers
+
+import (
+	"bufio"
+	"context"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+	"fmt"
+	"net"
+	"syscall"
+
+ 
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	persistapi "github.com/kata-containers/runtime/virtcontainers/persist/api"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/kata-containers/runtime/virtcontainers/store"
+	"github.com/kata-containers/runtime/virtcontainers/types"
+	"github.com/kata-containers/runtime/virtcontainers/utils"
+	"github.com/kata-containers/runtime/virtcontainers/device/config"
+	
+	govmmIch "github.com/kata-containers/runtime/virtcontainers/pkg/intel/govmm/ich"
+	
+)
+
+//
+// Constants and type definitions related to cloud hypervisor
+//
+
+type ichState uint8
+
+const (
+	ichNotReady ichState = iota
+	ichReady
+)
+
+const (
+	ichTimeout            = 10
+	ichSocket             = "ich.sock"
+	vfsdSocket            = "virtiofsd.sock"
+	apiSocket             = "cloud-hypervisor.9166"
+	ichStopSandboxTimeout = 15
+)
+
+// the info struct goes to disk as hypervisor.json!
+
+type CloudHypervisorInfo struct {
+	PID int
+}
+
+//
+// cloud hypervisor state
+//
+type CloudHypervisorState struct {
+	sync.RWMutex
+	state ichState
+	PID int
```

Comment:
moved.

Reply:
I've removed the govmm dependency and added a new CLI builder into the driver itself as agreed. I've also merged the info and state struct to simplify things and checked that the state in hypervisor.json is correct at runtime. I've also noticed that IF the hypervisor fails to start for whatever reason the virtiofsd obviously not teared-down by the hypervisor as a consequence so I use now the stored virtiofsd PID to kill the it to avoid accumulating unused processes.